File: framework/test2/common/reverse/reverse.go
package main

import (
    . "missingjs.com/soda/leetcode"
    "missingjs.com/soda/unittest"
    "missingjs.com/soda/util"
)

var logger = util.Logger
var _used ListNode

func reverseVowels(s string) string {
    isv := make([]bool, 128)
    vs := "aeiouAEIOU"
    for _, ch := range vs {
        isv[ch] = true
    }

    buf := []byte(s)
    i, j := 0, len(s) - 1
    for i < j {
        for i < j && !isv[buf[i]] {
            i++
        }
        for i < j && !isv[buf[j]] {
            j--
        }
        if i < j {
            temp := buf[i]
            buf[i] = buf[j]
            buf[j] = temp
            i++
            j--
        }
    }
    return string(buf)
}

func main() {
    // create tester by function
    work := unittest.CreateWork(reverseVowels)
    // OR create by struct factory
    // work := unittest.CreateWorkForStruct(Constructor)
    // work.SetValidator(func(e,r)bool)
    work.CompareSerial = true
    work.Run()
}

File: framework/test2/common/reverse/reverse.php
<?php
require_once 'Soda/DS/index.php';
require_once 'Soda/Unittest/index.php';

use Soda\DS;
use Soda\Unittest\TestWork;
use Soda\Unittest\Utils;

class Solution {

    /**
     * @param String $s
     * @return String
     */
    function reverseVowels($s) {
        $isv = array_fill(0, 128, false);
        $vowels = "aeiouAEIOU";
        for ($i = 0; $i < strlen($vowels); ++$i) {
            $ch = $vowels[$i];
            $isv[ord($ch)] = true;
        }
        $buf = str_split($s);
        [$i, $j] = [0, strlen($s)-1];
        while ($i < $j) {
            while ($i < $j && !$isv[ord($buf[$i])]) {
                ++$i;
            }
            while ($i < $j && !$isv[ord($buf[$j])]) {
                --$j;
            }
            if ($i < $j) {
                [$buf[$i], $buf[$j]] = [$buf[$j], $buf[$i]];
                ++$i;
                --$j;
            }
        }
        return join('', $buf);
    }
}

$work = TestWork::create(new Solution(), 'reverseVowels', $argv[0]);
// $work->validator = fn($e, $r) => <boolean result>
$work->compareSerial = true;
echo $work->run(Utils::fromStdin());
